numpy/_core/strings.py

```diff
@@ -608,33 +611,40 @@ def center(a, width, fillchar=' '):
     >>> np.strings.center(c, width=9, fillchar='*')
     array(['***a1b2**', '***1b2a**', '***b2a1**', '***2a1b**'], dtype='<U9')
     >>> np.strings.center(c, width=1)
-    array(['a', '1', 'b', '2'], dtype='<U1')
+    array(['a1b2', '1b2a', 'b2a1', '2a1b'], dtype='<U4')
 
     """
-    a_arr = np.asarray(a)
-    width_arr = np.asarray(width)
-    size = int(np.max(width_arr.flat))
-    if np.issubdtype(a_arr.dtype, np.bytes_):
-        fillchar = np._utils.asbytes(fillchar)
-    return _vec_string(
-        a_arr, type(a_arr.dtype)(size), 'center', (width_arr, fillchar))
+    a = np.asanyarray(a)
+    width = np.maximum(str_len(a), width)
+
+    fillchar = np.asanyarray(fillchar)
+    if np.any(str_len(fillchar) != 1):
+        raise TypeError(
+            "The fill character must be exactly one character long")
+
+    shape = np.broadcast_shapes(a.shape, width.shape, fillchar.shape)
+    if a.dtype.char == "T":
+        out = np.empty_like(a, shape=shape)
```

Comment:
I think I wondered about this before, but logically the ufunc machinery should do this already for `StringDType`.

p.s. Ah, yes, I wondered below, in https://github.com/numpy/numpy/pull/25908/files#r1509243153